src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -1872,162 +2022,323 @@ WorkerExtendedOpNode(MultiExtendedOp *originalOpNode,
 		 */
 		if (hasAggregates && !hasWindowFunction)
 		{
-			WorkerAggregateWalker((Node *) originalExpression, walkerContext);
+			WorkerAggregateWalker((Node *) originalExpression, workerAggContext);
 
-			newExpressionList = walkerContext->expressionList;
-			queryHasAggregates = true;
+			newExpressionList = workerAggContext->expressionList;
 		}
 		else
 		{
 			newExpressionList = list_make1(originalExpression);
 		}
 
-		createNewGroupByClause = walkerContext->createGroupByClause;
-
 		ProcessWorkerExpressionList(newExpressionList, originalTargetEntry,
-									createNewGroupByClause, &newTargetEntryList,
-									&targetProjectionNumber, &groupClauseList,
-									&nextSortGroupRefIndex);
+									workerAggContext->createGroupByClause,
+									newTargetEntryList, targetProjectionNumber,
+									groupClauseList, nextSortGroupRefIndex);
 	}
+}
+
+
+/*
+ * ProcessHavingClauseForWorkerQuery gets the inputs and modifies the outputs
+ * such that the worker query's target list and group by clauses are extended
+ * based on the inputs. The rationale is that Citus may need to apply the HAVING
+ * clause on the coordinator. Thus, the updated entries are intended to pull
+ * the necessary data from the workers.
+ *
+ * The function also updates the workerHavingQual if the having clause can
+ * be safely pushed down to the worker query.
+ *
+ * TODO: Do we really need to pull anything to the coordinator if we're already
+ * pushing down the HAVING clause?
+ *
+ *     inputs: originalHavingQual, extendedOpNodeStats
+ *     outputs: workerHavingQual, newTargetEntryList, targetProjectionNumber,
+ *              groupClauseList, nextSortGroupRefIndex
```

Comment:
since the function names are not that descriptive, I'm going to want descriptions for all the parameters.

Reply:
Are you OK with the format I propose here: https://github.com/citusdata/citus/pull/2090/files#diff-b067f8998f46f5afc9aad3fc1c35aa09R2060

Reply:
Actually, rebasing this PR on to the previous ones took more than I thought. 

Feel free to add comments here but I'll probably apply them after we merge the previous ones to keep the rebase efforts smaller.